mlir/tools/mlir-tblgen/OpDefinitionsGen.cpp

```diff
@@ -1143,6 +1143,8 @@ void OpEmitter::genAttrNameGetters() {
       const char *const getAttrName = R"(
   assert(index < {0} && "invalid attribute index");
   assert(name.getStringRef() == getOperationName() && "invalid operation name");
+  assert(!name.getAttributeNames().empty() && "empty attribute names. Is a new "
```

Comment:
I'm wondering if it is correct for `getAttributeNames` to return an empty list if the dialect is not loaded. The documentation of `getAttributeNames` talks about caching and maybe this should have fallen back to `OpTy::getAttributeNames`? Maybe @joker-eph knows.

Reply:
I was wondering about that too. The attribute names are all known at compile time, so you would guess that  `getFunctionTypeAttrName` (which figures out the name of the attribute `function_type`) can directly return the result. However, `getFunctionTypeAttrName` returns not only the name but also some context. I've tried passing the name as string instead of an `OperationName`, but that will crash since, if I remember correctly, the `FuncOp` could not be created since it was an unregistered op. So the fallback to (static?) `OpTy::getAttributeNames` is probably not gonna work.

Returning a `nullptr` for `getAttributeNames` sounds like something that may work.

Reply:
Since fixing this would require a separate PR anyway, I've opened an issue for this: https://github.com/llvm/llvm-project/issues/72900.

Reply:
I suspect a better assert here may be `assert(name.isRegistered() && "Operation isn't registered, missing a dependent dialect loading?");`

Reply:
Good tip. Thanks! I've implemented it and verified that it is triggered when reverting the fix from this PR (`func::FuncDialect` in `dependentDialects`).